lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -0,0 +1,156 @@
+
+/**
+ * Process directives.
+ *
+ * @param WP_Directive_Processor $tags An instance of the WP_Directive_Processor.
+ * @param string                 $prefix Attribute prefix.
+ * @param string[]               $directives Directives.
+ *
+ * @return WP_Directive_Processor The modified instance of the
+ * WP_Directive_Processor.
+ */
+function gutenberg_interactivity_process_directives( $tags, $prefix, $directives ) {
+	$context   = new WP_Directive_Context;
+	$tag_stack = array();
+
+	while ( $tags->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+		$tag_name = strtolower( $tags->get_tag() );
+
+		// Is this a tag that closes the latest opening tag?
+		if ( $tags->is_tag_closer() ) {
+			if ( 0 === count( $tag_stack ) ) {
+				continue;
+			}
+
+			list( $latest_opening_tag_name, $attributes ) = end( $tag_stack );
+			if ( $latest_opening_tag_name === $tag_name ) {
+				array_pop( $tag_stack );
+
+				// If the matching opening tag didn't have any attribute directives,
+				// we move on.
+				if ( 0 === count( $attributes ) ) {
+					continue;
+				}
+			}
+		} else {
+			// Helper that removes the part after the double hyphen before looking for
+			// the directive processor inside `$attribute_directives`.
+			$get_directive_type = function ( $attr ) {
+				return explode( '--', $attr )[0];
+			};
+
+			$attributes = $tags->get_attribute_names_with_prefix( $prefix );
+			$attributes = array_map( $get_directive_type, $attributes );
+			$attributes = array_intersect( $attributes, array_keys( $directives ) );
+
+			// If this is an open tag, and if it either has attribute directives, or
+			// if we're inside a tag that does, take note of this tag and its
+			// attribute directives so we can call its directive processor once we
+			// encounter the matching closing tag.
+			if (
+				! WP_Directive_Processor::is_html_void_element( $tags->get_tag() ) &&
+				( 0 !== count( $attributes ) || 0 !== count( $tag_stack ) )
+			) {
+				$tag_stack[] = array( $tag_name, $attributes );
+			}
+		}
+
+		foreach ( $attributes as $attribute ) {
+			call_user_func( $directives[ $attribute ], $tags, $context );
+		}
+	}
+
+	return $tags;
+}
+
+/**
+ * Resolve the reference using the store and the context from the provided path.
+ *
+ * @param string $path Path.
+ * @param array  $context Context data.
+ * @return mixed
+ */
+function gutenberg_interactivity_evaluate_reference( $path, array $context = array() ) {
+	$store = array_merge(
+		WP_Interactivity_Store::get_data(),
+		array( 'context' => $context )
+	);
+
+	if ( strpos( $path, '!' ) === 0 ) {
+		$path                  = substr( $path, 1 );
+		$has_negation_operator = true;
+	}
+
+	$array   = explode( '.', $path );
+	$current = $store;
+	foreach ( $array as $p ) {
+		if ( isset( $current[ $p ] ) ) {
+			$current = $current[ $p ];
+		} else {
+			return null;
+		}
+	}
+
+	// Check if $current is a function and if so, call it passing the store.
+	if ( is_callable( $current ) ) {
```

Comment:
note that the comment and code contradict each other here. almost anything can be a callable, including things that we don't want to treat as callable here. `"file"` is callable, because there's a PHP function named `file`. this means that it's not possible to use a context path like `users.avatar.file` because we will try and _run_ `file` and this will likely crash in surprising and unexpected ways.

given that the default behavior here is the dangerous and unsurprising one, it would be valuable to restrict this either to actual function data types or require more strict means for indicating that something _ought_ to be called.

Reply:
There are other callable types, like an array of two strings, e.g., `array( 'MyClass', 'some_static_method' )`. Those are harder to be evaluated as real callable, but it is another possibility. 🤔 

We could restrict `$current` to be an instance of `Closure`, which is the class that anonymous and arrow functions use. I don't think developers should be allowed to use a different callable type inside the store.

Reply:
Sounds good to me.

Reply:
And, in any case, if you want to use another callable, you can do so inside an anonymous function. 🤷